File: GKUserKitExample/PersistenStack.m
//
//  PersistenStack.m
//  NewTest
//
//  Created by 童小波 on 15/3/6.
//  Copyright (c) 2015年 tongxiaobo. All rights reserved.
//

#import "PersistenStack.h"

@interface PersistenStack()

@property (nonatomic,strong,readwrite) NSManagedObjectContext *managedObjectContext;
@property (nonatomic,strong) NSURL *modelURL;
@property (nonatomic,strong) NSURL *storeURL;

@end

@implementation PersistenStack

- (id)initWithStoreURL:(NSURL *)storeURL modelURL:(NSURL *)modelURL
{
    self = [super init];
    if (self) {
        self.storeURL = storeURL;
        self.modelURL = modelURL;
        [self setupManagedObjectContext];
    }
    return self;
}

- (void)setupManagedObjectContext
{
    self.managedObjectContext = [[NSManagedObjectContext alloc] initWithConcurrencyType:NSMainQueueConcurrencyType];
    self.managedObjectContext.persistentStoreCoordinator = [[NSPersistentStoreCoordinator alloc] initWithManagedObjectModel:self.managedObjectModel];
    NSError *error;
    [self.managedObjectContext.persistentStoreCoordinator addPersistentStoreWithType:NSSQLiteStoreType configuration:nil URL:self.storeURL options:nil error:&error];
    if (error) {
        NSLog(@"error: %@", error);
    }
    self.managedObjectContext.undoManager = [[NSUndoManager alloc] init];
}

- (NSManagedObjectModel *)managedObjectModel
{
    return [[NSManagedObjectModel alloc] initWithContentsOfURL:self.modelURL];
}

@end

File: GKUserKitExample/PersistenStack.swift
//
//  PersistenStack.swift
//  CoreDataDemo
//
//  Created by 童小波 on 15/1/14.
//  Copyright (c) 2015年 tongxiaobo. All rights reserved.
//

import UIKit
import CoreData

@objc class PersistenStack : NSObject {
  
  var managedObjectContext: NSManagedObjectContext?
  var backgroundManagedObjectContext: NSManagedObjectContext?
  var storeURL: NSURL!
  var modelURL: NSURL!
  
  init(storeURL: NSURL, modelURL: NSURL) {
    super.init()
    self.storeURL = storeURL
    self.modelURL = modelURL
    self.setupManagedObjectContexts()
  }
  
  func setupManagedObjectContexts(){
    managedObjectContext = self.setupManagedObjectContextWithConcurrencyType( NSManagedObjectContextConcurrencyType.MainQueueConcurrencyType)
    self.managedObjectContext?.undoManager = NSUndoManager()
    
    backgroundManagedObjectContext = self.setupManagedObjectContextWithConcurrencyType(NSManagedObjectContextConcurrencyType.PrivateQueueConcurrencyType)
    self.managedObjectContext?.undoManager = nil
    
    NSNotificationCenter.defaultCenter().addObserverForName(NSManagedObjectContextDidSaveNotification, object: nil, queue: nil) { (note) -> Void in
      let moc = self.managedObjectContext
      if note.object != nil{
        moc?.performBlock({ () -> Void in
          moc?.mergeChangesFromContextDidSaveNotification(note)
          return
        })
      }
    }
  }
  
  func setupManagedObjectContextWithConcurrencyType(concurrency :NSManagedObjectContextConcurrencyType) -> NSManagedObjectContext{
    let managedObjectContext = NSManagedObjectContext(concurrencyType: concurrency)
    managedObjectContext.persistentStoreCoordinator = NSPersistentStoreCoordinator(managedObjectModel: managedObjectModel())
    var error: NSError?
    managedObjectContext.persistentStoreCoordinator?.addPersistentStoreWithType(NSSQLiteStoreType, configuration: nil, URL: self.storeURL, options: nil, error: &error)
    if error != nil{
      println(error?.localizedDescription)
      println("rm \"\(self.storeURL.path))\"")
    }
    return managedObjectContext
  }
  
  func managedObjectModel() -> NSManagedObjectModel{
    return NSManagedObjectModel(contentsOfURL: self.modelURL)!
  }
  
}


